File: DP/2D/DP_on_Subsequence/parition_equal_subset_sum/memoized.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total_sum = sum(nums)
        if total_sum % 2 == 1:
            return False
        
        target = total_sum//2
        n = len(nums)

        def dp(i,target,memo):
            if(i,target) in memo:
                return memo[(i,target)]
            if target == 0:
                memo[(i,target)] = True
                return memo[(i,target)]
            if i == 0:
                if nums[0] == target:
                    memo[(i,target)] = True
                else:
                    memo[(i,target)] = False
                return memo[(i,target)]
            notTake = dp(i-1,target,memo)
            take = False
            if nums[i]<=target:
                take = dp(i-1,target-nums[i],memo)
            memo[(i,target)] = notTake or take
            return memo[(i,target)]


        res = dp(n-1,target,{})
        return res

File: DP/2D/DP_on_Subsequence/Coin_Change_MinimumCoin/memoized.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        Map<String,Integer> memo = new HashMap<>();
        int res = dp(coins.length - 1,amount,coins,memo);
        if (res >= (int) Math.pow(10,9)){
            return -1;
        }
        return res;
    }

    public int dp(int i,int amt,int[] coins, Map<String,Integer> memo){
        String key = i + "-" + amt;
        if(memo.containsKey(key)){
            return memo.get(key);
        }
        if (i == 0){
            if (amt % coins[0] == 0){
                return amt / coins[0];
            }else{
                return (int) Math.pow(10,9);
            }
        }

        int notTake = 0 + dp(i-1,amt,coins,memo);
        int take = (int) Math.pow(10,9);
        if (coins[i] <= amt){
            take = 1 + dp(i,amt - coins[i],coins,memo);
        }
        
        int res = Math.min(take,notTake);
        memo.put(key,res);
        return res;
    }
}